src/sema/expression/constructor.rs

```diff
@@ -577,29 +603,46 @@ pub(super) fn deprecated_constructor_arguments(
 
 /// When calling a constructor on Solana, we must verify it the contract we are instantiating has
 /// a program id annotation and require the accounts call argument if the call is inside a loop.
-fn solana_constructor_check(
+pub(super) fn solana_constructor_check(
     loc: &pt::Loc,
     constructor_contract_no: usize,
     diagnostics: &mut Diagnostics,
     context: &ExprContext,
     call_args: &CallArgs,
-    ns: &Namespace,
+    ns: &mut Namespace,
 ) {
-    if ns.target != Target::Solana {
-        return;
-    }
-
-    if ns.contracts[constructor_contract_no].program_id.is_none() {
+    if !ns.contracts[constructor_contract_no].instantiable {
         diagnostics.push(Diagnostic::error(
             *loc,
             format!(
-                "in order to instantiate contract '{}', a @program_id is required on contract '{}'",
+                "cannot construct '{}' of type '{}'",
                 ns.contracts[constructor_contract_no].name,
-                ns.contracts[constructor_contract_no].name
+                ns.contracts[constructor_contract_no].ty
             ),
         ));
     }
 
+    if let Some(context_contract) = context.contract_no {
```

Comment:
Not sure why this check is needed, and why it is not needed when called from a freestanding function.

Reply:
Good point. In fact, we do not allow constructors in standalone functions.

Having said that, I did not cover such a case in my syntax refactoring.